packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,157 @@
+import { Construct } from './construct';
+import { Token } from './tokens';
+
+/**
+ * ITaggable indicates a entity manages tags via the `tags` property
+ */
+export interface ITaggable {
+    tags: TagManager,
+}
+
+/**
+ * A single tag property values
+ */
+interface ITagProps {
+    value: string | Token;
+    propagate?: boolean;
+}
+
+/**
+ * A an object of tags
+ */
+export interface ITags {
+    [key: string]: ITagProps;
+}
+
+/**
+ * A single tag
+ */
+export interface ITag {
+    key: string;
+    value: string | Token;
+    propagate?: boolean;
+}
+
+/**
+ * TagManager facilitates a common implementation of tagging for Constructs.
+ *
+ * Each construct that wants to support tags should implement the `ITaggable`
+ * interface and properly pass tags to the `Resources` (Cloudformation) elements
+ * the `Construct` creates.
+ */
+export class TagManager {
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    private readonly tags: ITags = {};
+
+    constructor(readonly parent: Construct, initialTags: ITags = {}) {
+        this.tags = initialTags;
+
+    }
+
+    /**
+     * Converts the `tags` to a Token for use in lazy evaluation
+     */
+    public toToken(propgateOnly = false): Token {
```

Comment:
`propagateOnly` wasn't initially needed, but as I implemented the `VpcNetwork` I discovered that I needed somewhere to support passing just the tags that should propagate to `Resource`s that were not directly implementing `ITaggable`. If I didn't do this I needed to `addTag` at a specific time to ensure propagation worked as expected. I preferred this over the latter.

Reply:
I changed the way propagation works by having children go back up to parents. However, this function is still needed and I'm open to improvements if this is confusing.